cpp/src/gandiva/engine.cc

```diff
@@ -319,6 +319,10 @@ Status Engine::FinalizeModule() {
 
   // do the compilation
   execution_engine_->finalizeObject();
+  ARROW_RETURN_IF(
+      execution_engine_->hasError(),
+      Status::CodeGenError("Compilation failed: ", execution_engine_->getErrorMessage()));
```

Comment:
This change is not directly to this PR, but previously error will not be reported after calling `finalizedObject`, which makes troubleshooting difficult.

Reply:
It turns out this `hasError` API is only available in some (newer) version of LLVM, I reverted this piece of change